File: 0916-word-subsets/0916-word-subsets.cpp
class Solution {
public:
    bool check(vector<int>&freq1, vector<int>&freq2) {
        for(int i=0; i<26; i++) {
            if(freq1[i] < freq2[i]) return false;
        }
        return true;
    }
    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        vector<string>ans;
        vector<int>freq2(26, 0);
        for(string str:words2) {
            vector<int>temp(26, 0);
            for(char c:str) {
                temp[c - 'a']++;
                freq2[c - 'a'] = max(freq2[c - 'a'], temp[c - 'a']);
            }
        }
        for(string str:words1) {
            vector<int>freq1(26, 0);
            for(char c:str) {
                freq1[c - 'a']++;
            }
            if(check(freq1, freq2)) {
                ans.push_back(str);
            }
        }
        return ans;
    }
};

File: 0916-word-subsets/0916-word-subsets.java
class Solution {
    boolean subset(int[] freq1, int[] freq2) {
        for(int i=0; i<26; i++) {
            if(freq1[i] < freq2[i]) return false;
        }
        return true;
    }
    public List<String> wordSubsets(String[] words1, String[] words2) {
        List<String>ans = new ArrayList<>();
        int []freq2 = new int [26];
        for(String str:words2) {
            int []temp = new int[26];
            for(char c:str.toCharArray()) {
                temp[c - 'a']++;
                freq2[c - 'a'] = Math.max(freq2[c - 'a'], temp[c - 'a']);
            }
        }
        for(String str:words1) {
            int []freq1 = new int [26];
            for(char c:str.toCharArray()) freq1[c - 'a']++;
            if(subset(freq1, freq2)) {
                ans.add(str);
            }
        }
        return ans;
    }
}